src/main.rs

```diff
@@ -49,15 +49,95 @@ use std::io::ErrorKind;
 use std::os::unix::fs::PermissionsExt;
 use std::os::unix::net::{UnixListener, UnixStream};
 use std::path::Path;
-use std::thread;
+use std::time::Duration;
 
 use anyhow::{Context, Result};
+use crossbeam_channel as channel;
 use slog::{debug, error, o, Drain};
+use threadpool::ThreadPool;
 
 mod ffi;
 mod handlers;
 mod protocol;
 
+fn main() -> Result<()> {
+    const SOCKET_PATH: &str = "/var/run/nscd/socket";
+    const N_WORKERS: usize = 256;
+    const HANDOFF_TIMEOUT: Duration = Duration::from_secs(3);
+
+    ffi::disable_internal_nscd();
+
+    let decorator = slog_term::TermDecorator::new().build();
+    let drain = slog_term::FullFormat::new(decorator).build().fuse();
+    let drain = slog_async::Async::new(drain).build().fuse();
+
+    let logger = slog::Logger::root(drain, slog::o!());
+
+    let (pool, handle) = worker_pool(logger.clone(), N_WORKERS);
+
+    let path = Path::new(SOCKET_PATH);
+    std::fs::create_dir_all(path.parent().expect("socket path has no parent"))?;
+    std::fs::remove_file(path).ok();
+    let listener = UnixListener::bind(path).context("could not bind to socket")?;
+    std::fs::set_permissions(path, std::fs::Permissions::from_mode(0o777))?;
+
+    for stream in listener.incoming() {
+        match stream {
+            Ok(stream) => {
+                // if something goes wrong and it's multiple seconds until we
+                // get a response, kill the process.
+                //
+                // the timeout here is set such that nss will fall back to system
+                // libc before this timeout is hit - clients will already be
+                // giving up and going elsewhere so crashing the process should
+                // not make a bad situation worse.
+                match handle.send_timeout(stream, HANDOFF_TIMEOUT) {
+                    Err(channel::SendTimeoutError::Timeout(_)) => {
+                        anyhow::bail!("timed out waiting for an available worker: exiting")
+                    }
+                    Err(channel::SendTimeoutError::Disconnected(_)) => {
+                        anyhow::bail!("aborting: worker channel is disconnected")
+                    }
+                    _ => { /*ok!*/ }
```

Comment:
Can we make the last clause just `Ok`? I'm worried about send_timeout starting to return new error types and us silently ignoring them. I'd rather have an exhaustive match block.